### ISSUE TITLE
Standardize twisted names examples to use task.react

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5989]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)| @wallrj reported|
|-|-|
|Trac ID|trac#5989|
|Type|enhancement|
|Created|2012-09-15 18:04:15Z|
|Branch|https://github.com/twisted/twisted/tree/examples-react-5989-3|

 * Use task.react: [#3270](https://github.com/twisted/twisted/issues/3270) introduced a new function task.react which runs a single function and shuts down the reactor reliably.

 * Tidy output: The results of example scripts should be neatly printed to stdout and certain expected errors should be printed to stderr.

 * Log uncaught exceptions: react takes care of logging any uncaught exceptions which in many examples are silently discarded ([#4008](https://github.com/twisted/twisted/issues/4008)) 

 * Standard usage docs: Also add the standard command line help and usage docs to the examples

 * Optional: Non zero exit code on errors: maybe postpone this until [#11335](https://github.com/twisted/twisted/issues/11335) is resolved.

 * Optional: Add unit tests for the examples: See [#11340](https://github.com/twisted/twisted/issues/11340) for ideas about how to do this.

See also:
 * [#5594](https://github.com/twisted/twisted/issues/5594), [#5596](https://github.com/twisted/twisted/issues/5596)

Attachments:

* [names-examples-5596-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1fd/1fddce6b252d9a996afdf789c3147025167128f9/d248030fc99dad049c106303a18d5428fff3563f.patch) (11744 bytes) - added by rwall on 2013-01-21 16:59:39Z - Use task.react in all names examples, standardise the script output and add some tests
* [examples-react-5989-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1fd/1fddce6b252d9a996afdf789c3147025167128f9/2642e49a23e540b86cf43132f34bdab91f5e9844.patch) (11793 bytes) - added by rwall on 2013-01-29 19:12:56Z - An updated patch without the tests for executability
* [examples-react-5989-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1fd/1fddce6b252d9a996afdf789c3147025167128f9/c03ba5e0c8169d4dcf624b5287719c7ecaa54f36.patch) (4051 bytes) - added by rwall on 2013-01-30 10:42:40Z - Update patch against source:branches/examples-react-5989
* [examples-react-5989-4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1fd/1fddce6b252d9a996afdf789c3147025167128f9/539d21db4405ef37591bd131ef3ec7a5f3478a35.patch) (9004 bytes) - added by rwall on 2013-02-21 17:28:04Z - Now with twisted.usage and other responses to the tom.prince revew.

<details><summary>Searchable metadata</summary>

```
trac-id__5989 5989
type__enhancement enhancement
reporter__rwall rwall
priority__normal normal
milestone__ 
branch__branches_examples_react_5989_3 branches/examples-react-5989-3
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1347732255000000 1347732255000000
changetime__1364602536000000 1364602536000000
version__None None
owner__ 
cc__rwall cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_1"></a>|@wallrj commented|
|-|-|

I added the suggested changes to a patch ([ticket:5596:gethostbyname-exit-5596-4.patch)](None/ticket:5596:gethostbyname-exit-5596-4.patch)) in ticket:5596#comment:6

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra commented|
|-|-|

(In [36994]) Branching to 'examples-react-5989'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra commented|
|-|-|

(In [36995]) apply names-examples-5596-1.patch. refs [#5989](https://github.com/twisted/twisted/issues/5989)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra commented|
|-|-|

Forced a [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/examples-react-5989).

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_7"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

Thanks thjs

I can see that my new tests are badly broken.

I'll remove the new example script permissions tests and upload another patch.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_8"></a>|@wallrj removed owner<br>@wallrj set status to `new`|
|-|-|

In [examples-react-5989-2.patch](None/examples-react-5989-2.patch)
 * I removed the executable mode tests - which were failing on buildbot and which were never going to work reliably on an svn checkout or in windows.
 * Fixed some pyflakes and pep8 warnings in the test module
 * renamed the test_helpUsage test.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_9"></a>|@wallrj commented|
|-|-|

Oh, and that patch is against trunk.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra commented|
|-|-|

Can you make a patch against the branch instead?

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_11"></a>|@wallrj commented|
|-|-|

Replying to [thijs](#note_10):
> Can you make a patch against the branch instead?

Sorry about that. I was in a rush and couldn't get bzr to behave.

See updated patch [examples-react-5989-3.patch](None/examples-react-5989-3.patch)

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_12"></a>|@tomprince set owner to @wallrj|
|-|-|

  1. I get [buildbot](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/examples-react-5989&num_builds=10) failures.
  1. I also get (different) errors doing `trial -u twisted.names.test.test_examples'
  1. The examples should use `twisted.python.usage`, instead of a using custom argument checking.
  1. It would be nice if there were some tests that actually exercise the code in the examples. As it is, `twisted.names.client`, `twisted.names.error` and `twisted.names.dns` could be emptied or otherwise completely changed and the tests would all still pass. (I guess this is probably beyond the scope of this ticket)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_13"></a>|@wallrj removed owner|
|-|-|

Replying to [tom.prince](#note_12):
>   1. I get
>   [https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/examples-react-5989&num_builds=10
>   buildbot] failures.


I think you must have started the build before applying my last patch.

The test failures are in the test_executable method, which I removed
in an earlier patch...

eg
twisted.names.test.test_examples.GetHostByNameTests.test_executable

...should no longer be in the branch.


>   1. I also get (different) errors doing `trial -u
>   twisted.names.test.test_examples'

Fixed. That was a problem with the ExampleTestCase class that I
imported from Divmod. There was an indentation problem and it was
reusing a variable name as a list and then as a FilePath.

>   1. The examples should use `twisted.python.usage`, instead of a
>   using custom argument checking.

Since none of the examples actually have an options (only positional
args) I was trying to keep it simple. But I guess its nice to demo the
twisted usage class. I updated the examples and as a result I also had
to update the various __docs__ for consistency with the twisted.usage
output format.

>   1. It would be nice if there were some tests that actually
>   exercise the code in the examples. As it is,
>   `twisted.names.client`, `twisted.names.error` and
>   `twisted.names.dns` could be emptied or otherwise completely
>   changed and the tests would all still pass. (I guess this is
>   probably beyond the scope of this ticket)

I haven't attempted this, but there is ticket [#11340](https://github.com/twisted/twisted/issues/11340) for better testing
of example scripts. That's where I read about the Divmod example
testing class.

See latest patch [examples-react-5989-4.patch](None/examples-react-5989-4.patch) which is
against the branch.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_14"></a>|@tomprince set owner to @wallrj|
|-|-|

A couple of points: 
  * twistedchecker error:
```
************* Module twisted.names.test.test_examples
W9013:139,0: Expected 3 blank lines, found 2
```
  * It doesn't appear to be documented, but `synopsis` should be a single line. I think `synopsis` should be the first line of the current docstring, and the rest should be the longdesc. Also, if you move the option definition up to the top, it would be reasonable to shorten the docstring to just a single line summarizing what the program does.

     (Note: I don't think there are any existing guidelines on how to write examples, so I'm trying to create them in my reviews. Feel free to disregard this, or make alternate suggestions.)

   * I'm aware of [#11340](https://github.com/twisted/twisted/issues/11340), but the scope of that is too big to handle all at once. It wants to be done for a single example, first. I'd suggest one of these, except they depend on the environment to work properly.

Please commit after addressing the first two points.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_15"></a>|@wallrj commented|
|-|-|

(In [37206]) Branching to 'examples-react-5989-2'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_16"></a>|@wallrj commented|
|-|-|

Replying to [tom.prince](#note_14):
> A couple of points:
>   * twistedchecker error:
> ```
> ************* Module twisted.names.test.test_examples
> W9013:139,0: Expected 3 blank lines, found 2
> ```

Fixed.

>   * It doesn't appear to be documented, but `synopsis` should be a
>   * single line. I think `synopsis` should be the first line of the
>   * current docstring, and the rest should be the longdesc. Also, if
>   * you move the option definition up to the top, it would be
>   * reasonable to shorten the docstring to just a single line
>   * summarizing what the program does.

Done. Well I left the longdesc in the docstring, but that gets used
automatically by usage.options

>      (Note: I don't think there are any existing guidelines on how
>      to write examples, so I'm trying to create them in my
>      reviews. Feel free to disregard this, or make alternate
>      suggestions.)

I agree. I think it looks better.
Also found a that you can do str(options) to get exactly the same
output as from --help. So I'm using that when there are option
validation errors.

>
>    * I'm aware of [#11340](https://github.com/twisted/twisted/issues/11340), but the scope of that is too big to handle
>    * all at once. It wants to be done for a single example,
>    * first. I'd suggest one of these, except they depend on the
>    * environment to work properly.

Yep.

>
> Please commit after addressing the first two points.

Also
 * added some test module references
 * fixed some whitespace issues.
 * added a news file.
 * merged forward and
   [rebuilt](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/examples-react-5989-2)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_17"></a>|@wallrj set status to `closed`|
|-|-|

(In [37209]) Merge examples-react-5989-2: Use task.react in twisted.names examples

Author: rwall
Reviewer: thijs, tom.prince
Fixes: [#5989](https://github.com/twisted/twisted/issues/5989)

 * Updated all twisted.names example scripts to use task.react, which
   shuts down the reactor reliably and logs uncaught exceptions.

 * Tidied up the output of the twisted.names example scripts. Results
   of all scripts are now neatly printed to stdout and certain
   expected errors are printed to stderr.

 * Standardized the example script --help output

 * Examples now return non-zero exit code on certain errors

 * Added some new unit tests for the twisted.names examples - testing
   for consistent usage and error messages and for a standard shebang
   line.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_18"></a>|@wallrj set status to `reopened`|
|-|-|

(In [37593]) Revert r37209: ExampleTestBase monkeypatches sys.stdout but doesn't replace it afterwards

```
11:55 < glyph> The example tests replace sys.stdout and don't replace it, which seems bad.
11:55 < glyph> twisted.names.test.test_examples.ExampleTestBase.setUp specifically
11:55 < exarkun> That _does_ maybe seem like it could cause a problem or two.
11:55 < glyph> Then, a logging test (just one!) fails
12:08 < rwall> glyph: That's my fault. I'll raise a new ticket and fix it.
```

Reopens: [#5989](https://github.com/twisted/twisted/issues/5989)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_19"></a>|@wallrj commented|
|-|-|

(In [37625]) Branching to 'examples-react-5989-3'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_20"></a>|@wallrj removed owner<br>@wallrj set status to `new`|
|-|-|

Ready for another review in log:branches/examples-react-5989-3
 * Now using TestCase.patch instead of manually monkey patching
   sys.stdout and sys.stderr - which ensures that sys is returned to
   its previous state after tests.
 * [Build results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/examples-react-5989-3)

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_21"></a>|@tomprince set status to `closed`|
|-|-|

(In [37878]) Merge examples-react-5989-3: Use task.react in twisted.names examples

Author: rwall
Reviewer: thijs, tom.prince
Fixes: [#5989](https://github.com/twisted/twisted/issues/5989)

 * Updated all twisted.names example scripts to use task.react, which
   shuts down the reactor reliably and logs uncaught exceptions.

 * Tidied up the output of the twisted.names example scripts. Results
   of all scripts are now neatly printed to stdout and certain
   expected errors are printed to stderr.

 * Standardized the example script --help output

 * Examples now return non-zero exit code on certain errors

 * Added some new unit tests for the twisted.names examples - testing
   for consistent usage and error messages and for a standard shebang
   line.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_22"></a>|@tomprince commented|
|-|-|

Hmm. I guess I should have reviewed this, and not merged it. In any case, the change looks correct, and the test are now passing on the machine where they were failing:
```
<cyli> tomprince: blargh, I forgot to hit enter.  sorry for the delay just checked it out, bin/trial twisted passes.
<cyli> tomprince: python3 tests also pass on my machine for that branch
```
